controls/DropDown.cpp

```diff
@@ -0,0 +1,282 @@
+
+See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+*/
+
+#include "BaseMenu.h"
+#include "DropDown.h"
+
+CMenuDropDown::CMenuDropDown() : BaseClass()
+{
+	SetSize(120, 35);
+	SetCharSize( QM_BOLDFONT );
+
+	eTextAlignment = QM_LEFT;
+	iFlags |= QMF_DROPSHADOW;
+
+	bDropUp = false;
+
+	m_iState = 0;
+	m_iCount = 0;
+	memset( m_szNames, 0, sizeof( m_szNames ) );
+}
+
+void CMenuDropDown::VidInit()
+{
+	iSelectColor.SetDefault( uiPromptTextColor );
+	iBackgroundColor.SetDefault( uiColorBlack );
+	iFgTextColor.SetDefault( uiInputFgColor );
+	iBgTextColor.SetDefault( uiPromptTextColor );
+
+	BaseClass::VidInit();
+
+	m_scItemSize = m_scSize;
+
+	if( bDropUp )
+	{
+		m_ArrowClosed.Load(UI_UPARROW);
+		m_ArrowOpened.Load(UI_UPARROWFOCUS);
+	}
+	else
+	{
+		m_ArrowClosed.Load(UI_DOWNARROW);
+		m_ArrowOpened.Load(UI_DOWNARROWFOCUS);
+	}
+	m_ArrowSize = EngFuncs::PIC_Size( m_ArrowClosed );
+}
+
+int CMenuDropDown::IsNewStateByMouseClick()
+{
+	int state = m_iState;
+	Point pt(m_scPos);
+
+	if( !bDropUp )
+		pt.y += m_scItemSize.h;
+
+	for( int i = 0; i < m_iCount; i++ )
+	{
+		if( UI_CursorInRect( pt, m_scItemSize ) && !(iFlags & (QMF_GRAYED|QMF_INACTIVE)))
+		{
+			state = i;
+			break;
+		}
+		pt.y += m_scItemSize.h;
+	}
+
+	return state;
+}
+
+bool CMenuDropDown::KeyUp( int key )
+{
+	// if( UI::Key::IsLeftMouse( key ) && FBitSet( iFlags, QMF_HASMOUSEFOCUS ) )
+	// {
+	// 	MenuToggle( );
+	// 	return true;
+	// }
+
+	return false;
+}
+
+bool CMenuDropDown::KeyDown( int key )
+{
+	bool ret = false;
+	int state = m_iState;
+
+	if( UI::Key::IsLeftMouse( key ) && FBitSet( iFlags, QMF_HASMOUSEFOCUS ) )
+	{
+		if( m_isOpen )
+			state = IsNewStateByMouseClick( );
+
+		MenuToggle( );
+		ret = true;
+	}
+
+	// if( m_isOpen && UI::Key::IsLeftMouse( key ) && FBitSet( iFlags, QMF_HASMOUSEFOCUS ) )
+	// 	state = IsNewStateByMouseClick( );
+
+	if( !FBitSet( iFlags, QMF_MOUSEONLY ) )
+	{
+		if( UI::Key::IsDownArrow( key ) )
+			state += 1;
+		else if( UI::Key::IsUpArrow( key ) )
+			state -= 1;
+		else if( UI::Key::IsEnter( key ) )
+		{
+			MenuToggle( );
+			return true;
+		}
+	}
+
+	if( state != m_iState )
+	{
+		m_iState = state;
+
+		if( m_iState < 0 )
+			m_iState = m_iCount - 1;
+		else if( m_iState >= m_iCount )
+			m_iState = 0;
+
+		SetCvar();
+		_Event( QM_CHANGED );
+
+		ret = true;
+	}
+
+	return ret;
+}
+
+void CMenuDropDown::SetMenuOpen( bool isOpen )
+{
+	if( m_isOpen && bDropUp )
+		m_scPos.y += m_scSize.h - m_scItemSize.h;
+
+	m_isOpen = isOpen;
+
+	if( m_isOpen )
+	{
+		m_scSize.h = m_scItemSize.h * m_iCount;
+
+		if( bDropUp )
+			m_scPos.y -= m_scSize.h;
+
+		m_scSize.h += m_scItemSize.h;
+	}
+	else
+		m_scSize.h = m_scItemSize.h;
+}
+
+void CMenuDropDown::Draw()
+{
+	uint textflags = ETF_NO_WRAP;
```

Comment:
AFAIR, `uint` is non-standart extension.
Will be better to use `unsigned` or `uint32_t`.

Reply:
It isn't, we define `uint` in `xash3d_types.h`

Reply:
Ok